docs/source/future/core/nrunner.rst

```diff
@@ -151,34 +151,68 @@ create Runnables based on the user's input from the command line::
 Runner
 ~~~~~~
 
-A runner is an active entity that acts on the information of a
-runnable.  A runner will usually be tied to an specific kind of
-runnable, and will to able to act upon the specific information that
-runnable kind provides.
+A Runner, within the context of the N(ext)Runner architecture, is an
+active entity that acts on the information that a runnable contains,
+and quite simply, should be able to run what the Runnable describes.
 
-The following pseudo-code may help to illustrate that::
+A Runner will usually be tied to a specific ``kind`` of Runnable.
+That type of relationship (Runner is capable of running kind "foo"
+and Runnable is of the same kind "foo") is the expected mechanism that
+will be employed when selecting a Runner.
 
-  runnable_instance = create_runnable('exec', uri='/bin/sleep')
-  if runnable_instance.kind == 'exec':
-     runner = create_runner_exec(runnable_instance)
+A Runner can take different forms, depending on which layer one is
+interacting with.  At the lowest layer, a Runner may be a Python class
+that inherits from :class:`avocado.core.nrunner.BaseRunner`, and
+implements at least a matching constructor method, and a ``run()``
+method that should yield dictionary(ies) as result(s).
 
-A runner should produce status information on the progress of the
-execution of a runnable.  If the runnable produces interesting
-information, it should forward that along.  For instance, using the
-``exec`` runner example, it's helpful to start producing status
-that the process has been created and it's running as soon as
-possible.  These can be as simple as a sequence of::
+At a different level, a runner can take the form of an executable that
+follows the ``avocado-runner-$KIND`` naming pattern and conforms to a
+given interface/behavior, including accepting standardized command
+line arguments and producing standardized output.
+
+.. tip:: for a very basic example of the interface expected, refer to
+         ``selftests/functional/test_nrunner_interface.py`` on the
+         Avocado source code tree.
+
+Runner output
++++++++++++++
+
+A Runner should, if possible, produce status information on the
+progress of the execution of a Runnable.  While the Runner is
+executing what a Runnable describes, should it produce interesting
+information, the Runner should attempt to forward that along its
+generated status.
+
+For instance, using the ``exec`` Runner example, it's helpful to start
+producing status that the process has been created and it's running as
+soon as possible, even if no other output has been produced by the
+executable itself.  These can be as simple as a sequence of::
```

Comment:
We will certainly get to the point where a minimal set of keys/values to be required from every compatible Runner will be very well defined, documented and tested.  This is what I meant by saying that a Runner must comply with an interface on both input and output. 

Let me also say that my initial understanding is that a Runner should carry *less* of the weight of providing information than the entity managing the Runner(s) (normally a Job).  Because Runner implementations will probably be more in number than say a Job implementation, it makes sense IMO to move as much as possible of the responsibility out of the Runner.  Another reason is that one of the keys to the success to the N(ext)Runner is that it should be simple for developers wanting to integrate their custom test types, which probably means writing a Runner.  We should let them start small, and then evolve their implementations.

Finally, I'm aware that the current core probably doesn't corroborate to that, given that it's still too simple and feature limited.